README.md

[![Build Status](https://travis-ci.org/akosela/go-cs.svg?branch=master)](https://travis-ci.org/akosela/go-cs)
[![Go Report Card](https://goreportcard.com/badge/github.com/akosela/go-cs)](https://goreportcard.com/report/github.com/akosela/go-cs)
[![GoDoc](https://godoc.org/github.com/akosela/go-cs/cs?status.svg)](https://godoc.org/github.com/akosela/go-cs/cs)

## go-cs 

Concurrent ssh client

cs is a program for concurrently executing local or remote commands on
multiple hosts.  It is using OpenSSH for running remote commands.  Its
main purpose is to help automate and manage large network of hosts.  So
in essence this tool is very similar to Ansible, Chef or Puppet, but
much more simpler and faster in execution.  Speed and simplicity are its
main goals.

## Install

#### Go get

```
$ go get github.com/akosela/go-cs/cs
# ln -s $GOPATH/bin/cs /usr/bin/cs
```

#### Linux (deb)

```
$ curl -OL https://github.com/akosela/go-cs/releases/download/v1.1/go-cs_1.1-1_amd64.deb
# dpkg -i go-cs_1.1-1_amd64.deb
```

#### CRUX

Package

```
$ curl -OL https://github.com/akosela/go-cs/releases/download/v1.1/go-cs-1.1-1.pkg.tar.xz
# mv go-cs-1.1-1.pkg.tar.xz go-cs#1.1-1.pkg.tar.xz
# pkgadd go-cs#1.1-1.pkg.tar.xz
```

Port

```
$ httpup sync https://raw.githubusercontent.com/akosela/crux-ports/master/#go-cs go-cs
# cd go-cs
# pkgmk -i -d
```

#### Linux

```
$ curl -OL https://github.com/akosela/go-cs/releases/download/v1.1/go-cs-1.1.linux.tar.gz
$ tar xvf go-cs-1.1.linux.tar.gz
$ cd go-cs
$ gzip cs.1
# cp cs /usr/bin ; cp cs.1.gz /usr/share/man/man1
```

#### FreeBSD

Package

```
# pkg install go-cs
```

Port

```
# cd /usr/ports/net/go-cs
# make install clean
```

#### OpenBSD

Package

```
# pkg_add go-cs
```

Port

```
# cd /usr/ports/sysutils/go-cs
# make install clean
```

#### MacOS

```
$ curl -OL https://github.com/akosela/go-cs/releases/download/v1.1/go-cs-1.1.darwin.tar.gz
$ tar xvf go-cs-1.1.darwin.tar.gz
$ cd go-cs
$ gzip cs.1
# cp cs /opt/local/bin ; cp cs.1.gz /opt/local/share/man/man1
```

## Man page

```
CS(1)                      BSD General Commands Manual                   CS(1)

NAME
     cs - concurrent ssh client

SYNOPSIS
     cs [-emqrstuVv1] [-c file] [-cmd] [-cname] [-d file] [-dd] [-du path]
        [-f script.sh] [-h hosts_file] [-hba] [-i identity_file] [-io] [-ip]
        [-l login_name] [-lip] [-mx] [-nc] [-nmap] [-ns] [-o output_file]
	[-P port] [-p path] [-ping] [-soa] [-to timeout] [-top] [-tr] [-tri]
	[-uname] [-vm] [command] [[user@]host] ...

DESCRIPTION
     cs is a program for concurrently executing local or remote commands on
     multiple hosts.  It is using OpenSSH for running remote commands.  Its
     main purpose is to help automate and manage large network of hosts.

     The options are as follows:

     -c file
             Copy file to the remote machine.

     -cmd    Run an arbitrary local command concurrently on multiple hosts.

     -cname  Run a local DNS query of type CNAME.

     -d file
             Download file from the remote machine.  It will be saved in a
             directory named after the remote host only when you download from
             multiple servers.

     -dd     Prints basic hardware information for a specific host (sudo(8)
             and dmidecode(8) required).

     -du path
     	     Prints the list of largest files in a specific directory (sudo(8) 
	     required, units in M).

     -e      Prints hosts with errors only.

     -f script.sh
             Run a local shell script on the remote host.

     -h hosts_file
             Reads hostnames from the given hosts_file.  Lines in the
             hosts_file can include commented lines beginning with a `#' and
             only one host per line is allowed.

     -hba    Lists installed HBAs and their WWPNs (Linux only).

     -i identity_file
             Selects a file from which the identity (private key) for public
             key authentication is read.  The default is ~/.ssh/id_rsa.

     -io     Run remote iostat(1) on a specific host.

     -ip     Lists network interfaces.

     -l login_name
             Specifies the user to log in as on the remote machine.  This also
             may be specified on a per-host basis on the command line.
     
     -lip    Scans storage interconnects for new luns (Linux only).

     -m      Lists mounted filesystems on a specific host.

     -mx     Run a local DNS query of type MX.

     -nc     Tests specified port with nc(1).  Default is 22/tcp.

     -nmap   Scans host with nmap(1).

     -ns     Run a local DNS query of type NS.

     -o output_file
             Saves standard output and standard error to a file.

     -P port
             Port to connect to on the remote host.

     -p path
             Specifies remote or local path for files in a remote copy or
             download mode.

     -ping   Sends ICMP ECHO_REQUEST to network hosts.

     -q      Quiet mode.  Supresses verbose standard output from remote
             machines.  This mode reports success or failure only.

     -r      Recursively copy entire directories.  It follows symbolic links
             encountered in the tree traversal.

     -s      Sort output.

     -soa    Run a local DNS query of type SOA.

     -t      Force pseudo-tty allocation.	

     -to timeout
             Specifies the timeout (in seconds) used when connecting to the
             SSH server.  The default value is 4 seconds.

     -top    Run remote top(1) in batch mode on a specific host.

     -tr     Run local traceroute(8).

     -tri    Run local traceroute(8) using ICMP (local sudo(8) required).

     -uname  Prints remote system information including OS version.

     -u      Run remote uptime(1) on a specific host.

     -V      Displays the version number and exit.

     -v      Verbose mode.  Causes cs to print debugging messages from ssh(1)
             about its progress.  This is helpful in debugging connection,
             authentication, and configuration problems.  Multiple -v options
             increase the verbosity.  The maximum is 3.

     -vm     Run remote vmstat(8) on a specific host.

     -1      One line mode, useful for sorting output later.

AUTHENTICATION
     The default method for authentication is a public key authentication
     which serves its purpose when dealing with multiple hosts.  You can read
     more about public key authentication in ssh(1).

EXIT STATUS
     The cs utility exits 0 on success, and >0 if an error occurs.

EXAMPLES
     Run a series of commands on hosts foo and bar:

           $ cs 'uptime; uname -a' foo bar

     Run a command on multiple hosts specified in a hosts_file:

           $ cs -h hosts_file uptime

     Run a local shell script on multiple hosts:

           $ cs -f script.sh foo{1..100}

     Run a local shell script with sudo(8) on multiple hosts:

           $ cs -t -f script.sh foo{1..100}

     Copy file to multiple hosts using a specific remote path:

           $ cs -c file -h hosts_file -p /foo/bar

     Download file from host foo:~ to the current working directory:

           $ cs -d file foo

     Download recursively files from /foo/bar from multiple hosts to a speci-
     fic local path /tmp with subdirectories named after remote hosts:

           $ cs -r -d '/foo/bar/*' -h hosts_file -p /tmp

     Run a command on multiple hosts and sort the output:

           $ cs -1 -h hosts_file 'free -m | grep Swap' | sort -rnk4 | head

     Run local ping(1) on multiple hosts:

           $ cs -ping foo{1..100}

     Show name and size of all files matching the glob pattern:

           $ cs -du '/var/log/mes*' foo{1..100}

SEE ALSO
     scp(1), ssh(1), ssh_config(5), sudo(8)

AUTHORS
     Andy Kosela <akosela@andykosela.com>

BSD                              July 19, 2017                             BSD  
```


CHANGES
go-cs v1.1                Feb 21, 2023

  * Bugfix: fix -ping on Darwin.

go-cs v1.0                Feb 17, 2023

  * Bugfix: fix passing an identity to ssh.
  * Bugfix: fix '-uname' on non Red Hat systems.
  * Change: refactor some code.

go-cs v0.9                Jul 19, 2017

  * Bugfix: improve running sudo(8).
  * Change: update man page.
  * Feature: add '-hba, -io, -lip, -m' options.

go-cs v0.8                Jul 10, 2017

  * Bugfix: fix '-ping' on OpenBSD and NetBSD.
  * Bugfix: fix bugs in '-du, -top, -vm'.
  * Change: separate '-ip' from '-dd'.
  * Change: update man page.

go-cs v0.7                Jun 30, 2017

  * Change: refactor some functions.
  * Feature: add '-cmd, -cname, -dd, -du, -e, -mx, -nc, -nmap, -ns,
    -ping, -soa, -top, -tr, -tri, -u, -uname, -vm' options.

go-cs v0.6                Oct 10, 2016

  * Bugfix: new '-t' option to make sudo(8) happy.
  * Change: rename '-t' to '-to'.

go-cs v0.5                May 11, 2015

  * Bugfix: clean checkArgs() function.
  * Bugfix: fix '-d' when downloading same file from multiple servers.
    Now it creates subdirectories named after remote hosts.

go-cs v0.4                Apr 23, 2015

  * Change: '-c', '-d', '-f' now accept arguments.
  * Feature: update and expand man page.

go-cs v0.3                Dec 30, 2014

  * Change: rename '-v' to '-V'.
  * Feature: add '-v' option to display verbose mode.
  * Feature: add '-1' option to display one line mode.  It is especially
    helpful for sorting output later.

go-cs v0.2               Sept 29, 2014

  * Bugfix: fix bug in error routine.
  * Bugfix: clean run() function.
  * Bugfix: fix bug with login option.
  * Change: rename cs.go -> cs/cs.go.
  * Feature: add godoc comments.
  * Feature: automatically add host key to known_hosts file.
  * Feature: add '-v' option to display version.
  * Security: securely call ssh(1) by hardcoding its path.

go-cs v0.1                    Sept 8, 2014

  * The first public version.


cs/cs.go
// Copyright 2014 Andy Kosela.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Concurrent ssh client.
// cs is a program for concurrently executing local or remote commands
// on multiple hosts.  It is using OpenSSH for running remote commands.
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	echo	= "/bin/echo"
	du      = "/usr/bin/du"
	host    = "/usr/bin/host"
	hostn	= "/bin/hostname"
	ifconf 	= "/sbin/ifconfig"
	lspci	= "/sbin/lspci"
	mpath	= "/sbin/multipath"
	nc      = "/usr/bin/nc"
	scp     = "/usr/bin/scp"
	ssh     = "/usr/bin/ssh"
	top     = "/usr/bin/top"
	uptime  = "/usr/bin/uptime"
	timeFmt = "02-Jan-2006 15:04:05"
)

func createFile(path string) *os.File {
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND,
		0666)
	if err != nil {
		fmt.Println("cs:", err)
		os.Exit(1)
	}
	return file
}

func openFile(path string) *os.File {
	file, err := os.Open(path)
	if err != nil {
		fmt.Println("cs:", err)
		os.Exit(1)
	}
	return file
}

func readFile(file *os.File) []string {
	var s []string
	r := bufio.NewReader(file)
	for {
		line, _ := r.ReadString('\n')
		if strings.HasPrefix(line, "#") {
			continue
		} else if line == "" {
			break
		}
		s = append(s, line)
	}
	defer file.Close()
	return s
}

func checkArgs(nocmd int, copy, download, file, hostsfile string, tty *bool,
	argv []string) (string, []string) {
	var command string
	var hosts []string

	if nocmd == 1 {
		if hostsfile == "" {
			if len(argv) < 1 {
				flag.Usage()
			}
			command = ""
			hosts = argv[0:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
			command = ""
		}

		return command, hosts
	}

	if file != "" || copy != "" || download != "" {
		if hostsfile == "" {
			if len(argv) < 1 {
				flag.Usage()
			}
			hosts = argv[0:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
		}
		if file != "" {
			f := openFile(file)
			if *tty {
				c := strings.Join(readFile(f), "")
				command = "sudo sh <<'EOF'" + c + "EOF"
			} else {
				command = strings.Join(readFile(f), "")
			}
		}
	} else {
		if hostsfile == "" {
			if len(argv) < 2 {
				flag.Usage()
			}
			command = argv[0]
			hosts = argv[1:]
		} else {
			f := openFile(hostsfile)
			hosts = readFile(f)
			command = argv[0]
		}
	}

	return command, hosts
}

func exist(hostname, path string) string {
	if fi, err := os.Stat(path + "/" + hostname); err != nil {
		os.Mkdir(path+"/"+hostname, 0755)
		return path + "/" + hostname
	} else if fi.IsDir() {
		return path + "/" + hostname
	}
	os.Mkdir(path+"/"+hostname+".host", 0755)
	return path + "/" + hostname + ".host"
}

func run(command, hostname, id, login, path, port, timeout, copy, disku,
	download string, dd, cname, hba, io, ip, lcmd, lip, netcat, nmap, ns,
	mount, mx, one, png, recursive, soa, up, verbose1, verbose2, verbose3,
	top1, tr, tri, tty, uname, vm *bool, ddir int, f *os.File) string {

	hostname = strings.Trim(hostname, "\n")
	batchmode := "-oBatchMode=yes"
	strict := "-oStrictHostKeyChecking=no"
	tout := "-oConnectTimeout=" + timeout
	flag := "-"
	flag2 := ""

	if id != "" {
		id = "-i" + id
	} else {
		id = "-2"
	}

	if login != "" {
		login = "-l" + login
	} else {
		login = "-2"
	}

	if *verbose1 {
		flag = "-v"
	} else if *verbose2 {
		flag = "-vv"
	} else if *verbose3 {
		flag = "-vvv"
	}

	var cmd *exec.Cmd
	if copy != "" && *recursive {
		cmd = exec.Command(scp, login, id, flag+"r", "-P", port,
			batchmode, strict, tout, copy, hostname+":"+ path)
	} else if copy != "" {
		cmd = exec.Command(scp, login, id, "-P", port, batchmode,
			strict, tout, copy, hostname+":"+path)
	} else if download != "" && *recursive {
		if ddir == 1 {
			path = exist(hostname, path)
		}

		cmd = exec.Command(scp, login, id, flag+"r", "-P", port,
			batchmode, strict, tout, hostname+":"+download, path)
	} else if download != "" {
		if ddir == 1 {
			path = exist(hostname, path)
		}

		cmd = exec.Command(scp, login, id, "-P", port, batchmode,
			strict, tout, hostname+":"+download, path)
	} else if *cname {
		cmd = exec.Command(host, "-tcname", hostname)
	} else if *dd {
		c := "sudo dmidecode -s system-product-name " +
			"; sudo dmidecode -s system-serial-number"

		cmd = exec.Command(ssh, login, id, "-q", flag+"tt", "-p",
			port, batchmode, strict, tout, hostname, c)
	} else if disku != "" {
		c := "if [ `uname -s` == OpenBSD ]; then sudo " + du +
			" -akx " + disku + " |sort -rn 2>/dev/null |head -20" +
			"; else sudo " + du + " -amx " + disku +
			" |sort -rn 2>/dev/null |head -20; fi"

		cmd = exec.Command(ssh, login, id, "-q", flag+"tt", "-p",
			port, batchmode, strict, tout, hostname, c)
	} else if *hba {
		c := "if [ `uname -s` == Linux ]; then " + lspci + " |" +
			"grep -i fibre; grep -v zZzZ -H " +
			"/sys/class/fc_host/host*/port_name; else echo " +
			"'cs: your OS is not supported'; fi"

		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, c)
	} else if *io {
		c := "if [ `uname -s` == Linux ]; then iostat -mx |grep Dev " +
			"|awk '{OFS=\"\t\" ; print $1, $4, $5, $6, $7, $10, " +
			"$12, $9}'; iostat -mx |grep -e dm- -e sd -e hd | " +
			"sort -rnk10 2>/dev/null |head -20 | " +
			"awk '{OFS=\"\t\" ; print $1, $4, $5, $6, $7, $10, " +
			"$12, $9}'; elif [ `uname -s` == FreeBSD ]; then " +
			"iostat -x | grep qlen ; iostat -x |grep -v device | " +
			"sort -rnk9 2>/dev/null |head -20; " +
			"else echo 'cs: your OS is not supported'; fi"

		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, c)
	} else if *ip {
		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, ifconf)
	} else if *lip {
		c := "if [ `uname -s` == Linux ]; then for e in `ls " +
			"/sys/class/fc_host/host*/issue_lip`; do echo $e; " +
			"echo '1' >$e; done; for e in `ls " +
			"/sys/class/scsi_host/host*/scan`; do echo $e; echo " +
			"'- - -' >$e; done; echo '# multipath -F'; " + mpath +
			" -F; echo '# multipath'; " + mpath +
			"; else echo 'cs: your OS is not supported'; fi"
		command = "sudo sh <<'EOF'" + "\n" + c + "\n" + "EOF"

		cmd = exec.Command(ssh, login, id, "-q", flag+"tt", "-p",
			port, batchmode, strict, tout, hostname, command)
	} else if *mount {
		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, "mount")
	} else if *mx {
		cmd = exec.Command(host, "-tmx", hostname)
	} else if *ns {
		cmd = exec.Command(host, hostname)
	} else if *nmap {
		cmd = exec.Command("nmap", hostname)
	} else if *netcat {
		cmd = exec.Command(nc, "-w1", hostname, port)
	} else if *png {
		var c string
		if runtime.GOOS == "linux" {
			c = "ping -nc1 -s16 -W3 " + hostname + " |grep from"
		} else if runtime.GOOS == "freebsd" || runtime.GOOS == "darwin" {
			c = "ping -nc1 -s16 -W3000 " + hostname + " |grep from"
		} else {
			c = "ping -nc1 -s16 -w3 " + hostname + " |grep from"
		}
		cmd = exec.Command("/bin/sh", "-c", c)
	} else if *lcmd {
		scmd := strings.Split(command, " ")

		switch len(scmd) {
		case 1:
			cmd = exec.Command(scmd[0], hostname)
		case 2:
			cmd = exec.Command(scmd[0], scmd[1], hostname)
		case 3:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], hostname)
		case 4:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], scmd[3],
				hostname)
		case 5:
			cmd = exec.Command(scmd[0], scmd[1], scmd[2], scmd[3],
				scmd[4], hostname)
		}
	} else if *soa {
		cmd = exec.Command(host, "-tsoa", hostname)
	} else if *top1 {
		c := "if [ `uname -s` == Linux ]; then " + top +
			" -cbn1 |grep -ve '\\[' -e " + top + " 2>/dev/null" +
			" |head -20; else " + top + " -b 13; fi"

		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, c)
	} else if *tr {
		cmd = exec.Command("traceroute", hostname)
	} else if *tri {
		cmd = exec.Command("sudo", "traceroute", "-I", hostname)
	} else if *uname {
		c := "if [ `uname -s` == Linux ]; then uname -a; " +
			"cat /etc/redhat-release 2>/dev/null; else uname -a; fi"

		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, c)
	} else if *up {
		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, uptime)
	} else if *vm {
		c := "if [ `uname -s` == Linux ]; then vmstat -SM; " +
			"elif [ `uname -s` == FreeBSD ]; then vmstat -h; " +
			"else vmstat; fi"

		cmd = exec.Command(ssh, login, id, "-p", port, batchmode,
			strict, tout, hostname, c)
	} else {
		if *tty {
			flag2 = "tt"
		}

		cmd = exec.Command(ssh, login, id, "-q", flag+flag2+"p",
			port, batchmode, strict, tout, hostname, command)
	}

	buf, err := cmd.CombinedOutput()
	if err != nil {
		return hostname + ": " + string(buf)
	}

	if *one {
		return hostname + ": " + string(buf)
	}
	return hostname + ":\n" + string(buf)
}

func main() {
	flag.Usage = func() {
		fmt.Println(
`usage: cs [-emqrstuVv1] [-c file] [-cmd] [-cname] [-d file] [-dd] [-du path]
	  [-f script.sh] [-h hosts_file] [-hba] [-i identity_file] [-io] [-ip]
	  [-l login_name] [-lip] [-mx] [-nc] [-nmap] [-ns] [-o output_file]
	  [-P port] [-p path] [-ping] [-soa] [-to timeout] [-top] [-tr]
	  [-tri] [-uname] [-vm] [command] [[user@]host] ...`)
		os.Exit(1)
	}

	lcmd := flag.Bool("cmd", false, "Local command")
	cname := flag.Bool("cname", false, "CNAME")
	copy := flag.String("c", "", "Copy")
	dd := flag.Bool("dd", false, "Dmidecode")
	disku := flag.String("du", "", "Du")
	download := flag.String("d", "", "Download")
	error := flag.Bool("e", false, "Error")
	file := flag.String("f", "", "Script file")
	hba := flag.Bool("hba", false, "HBA")
	hostsfile := flag.String("h", "", "Hosts file")
	id := flag.String("i", "", "Identity file")
	io := flag.Bool("io", false, "Iostat")
	ip := flag.Bool("ip", false, "IP")
	lip := flag.Bool("lip", false, "LIP")
	login := flag.String("l", "", "Login name")
	mount := flag.Bool("m", false, "Mounts")
	mx := flag.Bool("mx", false, "MX")
	netcat := flag.Bool("nc", false, "Netcat")
	nmap := flag.Bool("nmap", false, "Nmap")
	ns := flag.Bool("ns", false, "NS")
	one := flag.Bool("1", false, "One line")
	out := flag.String("o", "", "Output filename")
	path := flag.String("p", ".", "Path")
	png := flag.Bool("ping", false, "Ping")
	port := flag.String("P", "22", "SSH port")
	quiet := flag.Bool("q", false, "Quiet")
	recursive := flag.Bool("r", false, "Recursive")
	soa := flag.Bool("soa", false, "SOA")
	sorted := flag.Bool("s", false, "Sort")
	timeout := flag.String("to", "4", "Timeout")
	top1 := flag.Bool("top", false, "Top")
	tr := flag.Bool("tr", false, "Traceroute")
	tri := flag.Bool("tri", false, "Traceroute -I")
	tty := flag.Bool("t", false, "Force pseudo-tty allocation")
	uname := flag.Bool("uname", false, "Uname")
	up := flag.Bool("u", false, "Uptime")
	version := flag.Bool("V", false, "Version")
	verbose1 := flag.Bool("v", false, "Verbose mode 1")
	verbose2 := flag.Bool("vv", false, "Verbose mode 2")
	verbose3 := flag.Bool("vvv", false, "Verbose mode 3")
	vm := flag.Bool("vm", false, "Vmstat")
	flag.Parse()
	argv := flag.Args()

	if *version {
		fmt.Println("cs 1.1")
		os.Exit(1)
	}

	nocmd := 0
	if *cname || *dd || *disku != "" || *hba || *io || *ip || *lip ||
		*mount || *netcat || *nmap || *ns || *mx || *png || *soa ||
		*top1 || *tr || *tri || *uname || *up || *vm {
		nocmd = 1
	}

	command, hosts := checkArgs(nocmd, *copy, *download, *file, *hostsfile,
		tty, argv)

	ddir := 0
	if len(hosts) > 1 {
		ddir = 1
	}

	var f *os.File
	if *out != "" {
		now := time.Now()
		nowStr := now.Format(timeFmt)
		f = createFile(*out)
		f.WriteString("------ START: " + nowStr + " ------\n")
	}

	var mk, mk2 []string
	if *sorted {
		mk = make([]string, len(hosts))
		if *quiet {
			mk2 = make([]string, len(hosts))
		}
	}

	output := make(chan string)
	for _, hostname := range hosts {
		go func(hostname string) {
			output <- run(command, hostname, *id, *login, *path,
				*port, *timeout, *copy, *disku, *download, dd,
				cname, hba, io, ip, lcmd, lip, netcat, nmap, ns,
				mount, mx, one, png, recursive, soa, up,
				verbose1, verbose2, verbose3, top1, tr, tri,
				tty, uname, vm, ddir, f)
		}(hostname)
	}

	e, err := 0, 0
	for i := 0; i < len(hosts); i++ {
		select {
		case c := <-output:
			err = 0
			if !strings.Contains(c, ":\n") {
				e++
				err = 1
				if !strings.Contains(c, "\n") {
					c = c + "\n"
				}
			}

			if *sorted {
				if *error && err == 0 {
					continue
				}
				if *quiet {
					if *out != "" {
						mk[i] = c
					}
					split := strings.Split(c, ":")
					if err == 1 {
						mk2[i] = split[0] +"\t[ERROR]\n"
						continue
					}
					mk2[i] = split[0] + "\t[OK]\n"
					continue
				}
				mk[i] = c
				continue
			}
			if *out != "" {
				if *error && err == 0 {
					continue
				}
				f.WriteString(c)
			}
			if *quiet {
				split := strings.Split(c, ":")
				if *error && err == 0 {
					continue
				}

				if err == 1 {
					fmt.Println(split[0] + "\t[ERROR]")
					continue
				}
				fmt.Println(split[0] + "\t[OK]")
				continue
			}

			if *error && err == 0 {
				continue
			}
			fmt.Print(c)
		}
	}

	if *sorted {
		sort.Strings(mk)
		if *out != "" {
			for _, v := range mk {
				f.WriteString(v)
			}
		}
		if *quiet {
			sort.Strings(mk2)
			for _, v := range mk2 {
				fmt.Print(v)
			}
		} else {
			for _, v := range mk {
				fmt.Print(v)
			}
		}
	}

	if !*one {
		fmt.Printf("hosts = %d, errors = %d\n", len(hosts), e)
	}
	if *out != "" {
		now := time.Now()
		nowStr := now.Format(timeFmt)
		f.WriteString("------   END: " + nowStr + " ------\n")
		f.Close()
	}
}
